### PR TITLE
[JENKINS-27505] Preserve an empty line at the beginning of a textarea

### PR DESCRIPTION
**Pull request by ikedam**
[JENKINS-27505](https://issues.jenkins-ci.org/browse/JENKINS-27505)
A blank line in the beginning of a textarea is removed.
It is caused for the specification of HTML:
http://www.w3.org/TR/html5/syntax.html#syntax-newlines

> A single newline may be placed immediately after the start tag of pre and textarea elements. If the element's contents are intended to start with a newline, two consecutive newlines thus need to be included by the author.

(I could not found the description in HTML4.01)

I think my change gets too complicated as I don't know the best way to handle `\r` and `\n` in jelly files
(jelly doesn't seem to handle backslashes).
I want help of an expert of jelly.

### CODE REVIEW

**Comment by ikedam** [comment]
This result demonstrates what happens:
https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/2256/testReport/lib.form/TextAreaTest/testTextBeginningWithEmptyLine/
```
org.junit.ComparisonFailure: expected:<[
]begin

with
empty
li...> but was:<[]begin

with
empty
li...>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.junit.Assert$assertEquals.callStatic(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:50)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:157)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:169)
	at lib.form.TextAreaTest.testTextBeginningWithEmptyLine(TextAreaTest.groovy:94)
```

I'll commit the fix.

File: core/src/main/resources/lib/form/textarea.jelly [hunk 1]
```diff
@@ -89,6 +89,7 @@ THE SOFTWARE.
             readonly="${attrs.readonly}"
             codemirror-mode="${attrs['codemirror-mode']}"
             codemirror-config="${attrs['codemirror-config']}">
+    <j:if test="${value != null &amp;&amp; !empty(value.toString()) &amp;&amp; (value.toString().codePointAt(0) == 10 || value.toString().codePointAt(0) == 13)}"><j:whitespace>&#10;</j:whitespace></j:if>
```
Comment (ikedam) [inline core/src/main/resources/lib/form/textarea.jelly]: Jelly doesn't accept backslashes (like `\n`)...

Comment (KostyaSha) [inline core/src/main/resources/lib/form/textarea.jelly]: jelly and groovy has ability to unescape text, feel free to try.

File: test/src/test/groovy/lib/form/TextAreaTest.groovy [hunk 2]
```diff
@@ -70,4 +72,50 @@ class TextAreaTest {
...
     }
+    
+    @Test
+    public void testText() {
+        def TEXT_TO_TEST = "some\nvalue\n";
+        def p = j.createFreeStyleProject();
+        p.buildersList.add(new TextareaTestBuilder(TEXT_TO_TEST));
+        assertEquals(TEXT_TO_TEST, p.getBuildersList().get(TextareaTestBuilder.class).getText());
+        j.configRoundtrip(p);
+        assertEquals(TEXT_TO_TEST, p.getBuildersList().get(TextareaTestBuilder.class).getText());
```
Comment (jglick) [inline test/src/test/groovy/lib/form/TextAreaTest.groovy]: Can use `assertEqualDataBoundBeans` to make it a bit more concise if you like.

File: test/src/test/groovy/lib/form/TextAreaTest.groovy [hunk 3]
```diff
@@ -70,4 +72,50 @@ class TextAreaTest {
...
+    }
+
+    @Issue("JENKINS-27505")
+    @Test
+    public void testTextBeginningWithEmptyLine() {
+        def TEXT_TO_TEST = "\nbegin\n\nwith\nempty\nline\n\n";
```
Comment (jglick) [inline test/src/test/groovy/lib/form/TextAreaTest.groovy]: What about _two_ initial newlines?

**Comment by jglick** [comment]
:+1:

File: core/src/main/resources/lib/form/textarea.jelly (hunk 1 above)
Comment (ikedam) [inline core/src/main/resources/lib/form/textarea.jelly]: I tried this in jelly
```
:\r\n
:${'\r\n'}
:${&quot;\r\n&quot;}
:<st:out value="\r\n" />
:<st:out value="${'\r\n'}" />
:<st:out value="${&quot;\r\n&quot;}" />
```
and that resulted like this:
```
:\r\n
:\r\n
:\r\n
:\r\n
:\r\n
:\r\n
```
backslashes doesn't seem handled.

Maybe I need another way to handle control characters.
Would you show me some examples to have jelly unescape texts?
I could not find good documentations for jelly...

**Comment by ikedam** [comment]
Updated tests:
* Added a test for two initial new lines.
* Now uses `assertEqualDataBoundBeans`. That's really useful.

File: core/src/main/resources/lib/form/textarea.jelly (hunk 1 above)
Comment (jglick) [inline core/src/main/resources/lib/form/textarea.jelly]: [Documentation for Jelly](http://commons.apache.org/proper/commons-jelly/tags.html#jelly:core) is weak. Generally speaking if you are trying to do anything subtle you have to switch to Groovy to get more control.

File: test/src/test/groovy/lib/form/TextAreaTest.groovy (hunk 3 above)
Comment (jglick) [inline test/src/test/groovy/lib/form/TextAreaTest.groovy]: (resolved)

File: test/src/test/groovy/lib/form/TextAreaTest.groovy [hunk 4]
```diff
@@ -70,4 +72,61 @@ class TextAreaTest {
...
+        def TEXT_TO_TEST = "some\nvalue\n";
+        def p = j.createFreeStyleProject();
+        def target = new TextareaTestBuilder(TEXT_TO_TEST);
+        p.buildersList.add(target);
+        j.configRoundtrip(p);
+        j.assertEqualDataBoundBeans(target, p.getBuildersList().get(TextareaTestBuilder.class));
+    }
+
+    @Issue("JENKINS-27505")
+    @Test
```
Comment (jglick) [inline test/src/test/groovy/lib/form/TextAreaTest.groovy]: BTW it would save test execution time to collapse these three test cases.

**Comment by jglick** [comment]
:+1:

**Comment by ikedam** [comment]
Collapsed test cases.

**Comment by oleg-nenashev** [comment]
:+1:

**Comment by oleg-nenashev** [comment]
I'm merging this implementation, because it has got more votes. #1643 may be considered as a future improvement

File: test/src/test/groovy/lib/form/TextAreaTest.groovy [hunk 5]
```diff
@@ -70,4 +72,62 @@ class TextAreaTest {
         }
 
     }
+    
+    @Issue("JENKINS-27505")
+    @Test
+    public void testText() {
+        T1:{
+            def TEXT_TO_TEST = "some\nvalue\n";
```
Comment (lanwen) [inline test/src/test/groovy/lib/form/TextAreaTest.groovy]: why not just create 3 test methods?

**Comment by lanwen** [comment]
@ikedam You can use `@ClassRule` for jenkins and separate test class. Collapsing tests is bad pattern with produces test smell

**Pull request merged**